### ISSUE TITLE
TPOTClassifier with custom cv parameter

### ISSUE THREAD

**Issue by avgaydashenko** [EpistasisLab/tpot#557]
Hi,

I would like to use TPOTClassifier with _my generator function_ as cv parameter. But I have some troubles even with the code provided in the answer here: https://github.com/rhiever/tpot/issues/551

So, here's the error:

> There was an error in the TPOT optimization process. This could be because the data was not formatted properly, or because data for a regression problem was provided to the TPOTClassifier object. Please make sure you passed the data to TPOT correctly.

![default](https://user-images.githubusercontent.com/11502139/29926948-87dc8aea-8e6d-11e7-9f3d-6b60c274dcc9.png)

Basically, my question is: what format my function should have?

**Comment by rhiever** [comment]
The error message you're seeing is a generic error message, likely because the `cross_val_score` function used inside TPOT is throwing errors. Looking at the code snippet you shared, the first thing to check is the format of the scoring function `my_score`. Does it follow the format shown in the example from the [TPOT scores section](http://rhiever.github.io/tpot/using/#scoring-functions)?

```Python
def my_custom_accuracy(y_true, y_pred):
    return float(sum(y_pred == y_true)) / len(y_true)
```

Essentially, the function must take `y_true` (array of numbers; the true labels for the data) and `y_pred` (array of numbers; the predicted labels from the pipeline) and return a single float value.

**Comment by avgaydashenko** [comment]
Here's my function (seems right to me):

```
def my_score(y_true, y_pred):
    res = 0
    for cl in range(0, 3):
        ind = y_true == cl
        res += (y_true[ind] == y_pred[ind]).sum() / ind.sum()
    return res / 3
```

**Comment by rhiever** [comment]
I think I spotted a bug in `my_score`: If `0`, `1`, or `2` aren't in `y_true`, then the next line will cause the function to return `nan`. For example, call the following:

```Python
def my_score(y_true, y_pred):
    res = 0
    for cl in range(0, 3):
        ind = y_true == cl
        res += (y_true[ind] == y_pred[ind]).sum() / ind.sum()
    return res / 3

print(my_score(np.array([1,2,3]), np.array([1,2,3])))
```

If you expect classes 0, 1, and 2 to always appear in the labels, it would be a good idea to put a check in the scoring function to make sure they're all there. If it's meant to be a genetic multi-class function, then [TPOT's balanced accuracy function](https://github.com/rhiever/tpot/blob/1b012a9ae5722ba31869ee556054e0f219424885/tpot/metrics.py#L45:L62) can make a good template for a scoring function that works with arbitrary class labels.

**Comment by avgaydashenko** [comment]
I rewrote my function as follows, still same result:

```
def my_score(y_true, y_pred):
    res = 0
    cl_num = 3
    for cl in range(0, 3):
        ind = y_true == cl
        if len(ind) == 0:
            cl_num -= 1
            continue
        res += (y_true[ind] == y_pred[ind]).sum() / ind.sum()
    return res / cl_num
```

**Comment by rhiever** [comment]
Okay, the next thing to check is the shape of the data. What is the output of the following?

```Python
print(X_train.shape, y_train.shape)
```

Below is a working example of TPOT applied to the iris multiclass problem with LeaveOneOut cv and `my_score` scoring function.

```Python
from tpot import TPOTClassifier
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split, LeaveOneOut
import numpy as np

iris = load_iris()
X_train, X_test, y_train, y_test = train_test_split(iris.data.astype(np.float64),
    iris.target.astype(np.float64), train_size=0.75, test_size=0.25)

def my_score(y_true, y_pred):
    res = 0
    cl_num = 3
    for cl in range(0, 3):
        ind = y_true == cl
        if len(ind) == 0:
            cl_num -= 1
            continue
        res += (y_true[ind] == y_pred[ind]).sum() / ind.sum()
    return res / cl_num

tpot = TPOTClassifier(generations=5, population_size=50, verbosity=2,
                      cv=LeaveOneOut(), scoring=my_score)
tpot.fit(X_train, y_train)
print(tpot.score(X_test, y_test))
```

**Comment by avgaydashenko** [comment]
Yep, got results with LOO, thanks!

Then back to my initial question about custom generator: what format should I use?

**Comment by rhiever** [comment]
Great! To answer that question, it must follow the input/output scheme of the other scikit-learn CV operators. Assuming you want to define your own predefined splits, you can use scikit-learn's [PredefinedSplit](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.PredefinedSplit.html#sklearn.model_selection.PredefinedSplit).

**Comment by avgaydashenko** [comment]
Nice, will do that, thanks a lot :)
